# <font color=RED> ROMPEBRECHAS </font>

**PROBLEMA**:
¿Qué factores influyen en casos de violencia escolar a nivel nacional en niños de educación primaria en el período 2024-2025?


**UNIDAD DE ANALÍSIS DE CADA BASE DE DATOS:**
*   Lineal_3AP tiene como unidad de analísis **los colegios a nivel nacional**

*   Por otro lado, denuncias escolares tiene como **unidad de analísis a cada caso de denuncia o hecho de violencia escolar**

*   Finalmente la base de datos de padrón web tiene como unidad de analísis **a cada colegio o institución educativa**




**FUENTES DE DATOS USADAS:** censoeducativo2025: https://escale.minedu.gob.pe/uee/-/document_library_display/GMv7/view/10632985 denunciasviolencia:
https://siseve.minedu.gob.pe/Web/App/Mapa

Escale base de datos de colegio:
https://escale.minedu.gob.pe/web/inicio/padron-de-iiee

In [ ]:
!pip install dbfread

## <font color=yellow> 1. Tratamiento de datos — Lineal_3AP_1 </font>

### <font color=skyblue> 1.1 Importación de datos — Lineal_3AP_1 </font>

In [ ]:
from dbfread import DBF
import pandas as pd
import numpy as np

ruta = "Lineal_3AP_1.dbf"

tabla_f2 = DBF(ruta, encoding="latin-1", char_decode_errors="replace")

df_censo = pd.DataFrame(iter(tabla_f2))

# Convertir nombres de columnas a mayúsculas
df_censo.columns = [c.upper() for c in df_censo.columns]

display(df_censo.head(5))

,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,P104A,P105A,P105A_S_1,...,P136B_S_1,P136B_S_2,P136B_S_3,P136B_S_4,P136B_S_4O,NIV_MOD,GES_DEP,AREA_CENSO,CODGEO,CODOOII
0,1720556,0,852373,3AP,SIR ALEXANDER FLEMING,CERCADO,,NO,,,...,X,X,,,,B0,B4,1,040101,040001
1,0313890,0,073181,3AP,40513 SAN MIGUEL ARCANGEL,ALCA,ANEXO DE CAHUANA,NO,,,...,X,X,X,,,B0,A1,2,040802,040009
2,1495092,0,605028,3AP,33421,PILLCO MARCA,CAYHUAYNA,NO,,,...,,,,,,B0,A1,1,100111,100001
3,1579861,0,651751,3AP,SANTANDER DE PUENTE PIEDRA,PUENTE PIEDRA,SANTA PAULA,NO,,,...,,,,,,B0,B4,1,150125,150105
4,1529767,0,553166,3AP,COLEGIO SUPERIOR DE CIENCIAS NEWTON VIANNEY,YURA,,NO,,,...,,X,,,,B0,B4,1,040128,040001


In [ ]:
df_censo.columns

Index(['COD_MOD', 'ANEXO', 'CODLOCAL', 'NROCED', 'CEN_EDU', 'DISTRITO',
       'LOCALIDAD', 'P104A', 'P105A', 'P105A_S_1',
       ...
       'P136B_S_1', 'P136B_S_2', 'P136B_S_3', 'P136B_S_4', 'P136B_S_4O',
       'NIV_MOD', 'GES_DEP', 'AREA_CENSO', 'CODGEO', 'CODOOII'],
      dtype='object', length=193)

### <font color=skyblue> 1.2 Ajuste de columnas — Lineal_3AP_1 </font>

In [ ]:
import pandas as pd

# 1. Lista de variables solicitadas (ahora coinciden directamente en mayúsculas)
columnas_solicitadas = [
    'COD_MOD', 'ANEXO', 'CODLOCAL', 'NROCED', 'CEN_EDU', 'DISTRITO', 'LOCALIDAD',
    'CODGEO', 'NIV_MOD', 'GES_DEP',
    'P106A', 'P107A_NO', 'P112A', 'P113A', 'P114A', 'P116A', 'P118A_S_1',
    'P118A_S_3', 'P121A', 'P122A', 'P103B', 'P105B', 'P106B',
    'P109B_4', 'P109B_6', 'P110B_3', 'P114B_3', 'P114B_4', 'P114B_5',
    'P114B_6', 'P114B_7', 'P122B', 'P123B', 'P124B', 'P124B_SI', 'P124B_NO',
    'P125B', 'P126B_SI_1', 'P127B', 'P128B', 'P129B'
]

# 2. Filtrar el DataFrame directamente
# Usamos una lista de comprensión para asegurar que solo pedimos columnas que existen en df_censo
columnas_existentes = [c for c in columnas_solicitadas if c in df_censo.columns]

df_censo_final = df_censo[columnas_existentes].copy()

print(f"Dataset final creado con {len(df_censo_final.columns)} variables verificadas.")
display(df_censo_final.head())


Dataset final creado con 41 variables verificadas.


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO,NIV_MOD,GES_DEP,...,P122B,P123B,P124B,P124B_SI,P124B_NO,P125B,P126B_SI_1,P127B,P128B,P129B
0,1720556,0,852373,3AP,SIR ALEXANDER FLEMING,CERCADO,,040101,B0,B4,...,SI,SI,NO,,2,SI,,SI,SI,SI
1,0313890,0,073181,3AP,40513 SAN MIGUEL ARCANGEL,ALCA,ANEXO DE CAHUANA,040802,B0,A1,...,SI,SI,SI,0,,SI,,NO,SI,SI
2,1495092,0,605028,3AP,33421,PILLCO MARCA,CAYHUAYNA,100111,B0,A1,...,SI,SI,SI,0,,SI,1,SI,SI,SI
3,1579861,0,651751,3AP,SANTANDER DE PUENTE PIEDRA,PUENTE PIEDRA,SANTA PAULA,150125,B0,B4,...,NO,SI,SI,0,,SI,,NO,NO,SI
4,1529767,0,553166,3AP,COLEGIO SUPERIOR DE CIENCIAS NEWTON VIANNEY,YURA,,040128,B0,B4,...,NO,SI,SI,0,,SI,,NO,NO,SI


In [ ]:
print(f"Total de registros en df_censo: {len(df_censo_final)}")
print(f"Dimensiones (filas, columnas): {df_censo_final.shape}")

Total de registros en df_censo: 37354
Dimensiones (filas, columnas): (37354, 41)


In [ ]:
print("Tipos de datos en df_censo:")
display(df_censo_final.dtypes)

Tipos de datos en df_censo:


,0
COD_MOD,object
ANEXO,object
CODLOCAL,object
NROCED,object
CEN_EDU,object
DISTRITO,object
LOCALIDAD,object
CODGEO,object
NIV_MOD,object
GES_DEP,object


### <font color=skyblue> 1.3 Descripción de datos — Lineal_3AP_1 </font>

In [ ]:
print("Resumen de df_censo:")
display(df_censo_final.describe(include="all"))

Resumen de df_censo:


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO,NIV_MOD,GES_DEP,...,P122B,P123B,P124B,P124B_SI,P124B_NO,P125B,P126B_SI_1,P127B,P128B,P129B
count,37354,37354,37354,37354,37354,37354,37354,37354,37354,37354,...,37354,37354,37354,37354,37354,37354,37354,37354,37354,37354
unique,37354,1,37349,1,33317,2899,20115,1889,1,10,...,3,3,3,30,6,3,52,3,3,3
top,0403493,0,322232,3AP,PRIMARIA,SAN JUAN DE LURIGANCHO,,150132,B0,A1,...,SI,SI,SI,0,,SI,,NO,SI,SI
freq,1,37354,2,37354,51,453,1158,490,37354,29024,...,31898,32531,30709,27474,30698,31045,32721,20831,32517,33156


### <font color=skyblue> 1.4 Chequeo de faltantes y atípicos — Lineal_3AP_1 </font>

In [ ]:
display(df_censo_final.isnull().sum())

# También podemos ver el porcentaje de valores faltantes.
print("Porcentaje de valores faltantes en df_censo:")
display((df_censo_final.isnull().mean() * 100).round(2))

,0
COD_MOD,0
ANEXO,0
CODLOCAL,0
NROCED,0
CEN_EDU,0
DISTRITO,0
LOCALIDAD,0
CODGEO,0
NIV_MOD,0
GES_DEP,0


Porcentaje de valores faltantes en df_censo:


,0
COD_MOD,0.0
ANEXO,0.0
CODLOCAL,0.0
NROCED,0.0
CEN_EDU,0.0
DISTRITO,0.0
LOCALIDAD,0.0
CODGEO,0.0
NIV_MOD,0.0
GES_DEP,0.0


In [ ]:
print("Valores distintos por variable en df_censo:")
display(df_censo_final.nunique())

Valores distintos por variable en df_censo:


,0
COD_MOD,37354
ANEXO,1
CODLOCAL,37349
NROCED,1
CEN_EDU,33317
DISTRITO,2899
LOCALIDAD,20115
CODGEO,1889
NIV_MOD,1
GES_DEP,10


In [ ]:
print("Valores de la variable De las actividades que realiza con familias y comunidad, \n¿por lo menos una de ellas tiene como eje principal la Educación Sexual Integral (ESI)?\nrecibió en df_censo:")
display(df_censo_final["P116A"].value_counts(dropna=False)) #dropna=False para que también se muestren los valores faltantes

Valores de la variable De las actividades que realiza con familias y comunidad, 
¿por lo menos una de ellas tiene como eje principal la Educación Sexual Integral (ESI)?
recibió en df_censo:


,count
P116A,
SI,16482
NO,13158
,7714


### <font color=skyblue> 1.5 Resumen de datos — Lineal_3AP_1 </font>

In [ ]:
print("================RESUMEN DEL CENSO================")
resumen_censo = pd.DataFrame({
    "fuente": "Lineal_3AP_1.dbf",
    "variable": df_censo_final.columns,
    "tipo_python": df_censo_final.dtypes.values,
    "n_faltantes": df_censo_final.isnull().sum().values,
    "porcentaje_faltantes": (df_censo_final.isnull().mean() * 100).round(2).values,
    "valores_distintos": df_censo_final.nunique().values
})

display(resumen_censo)

================RESUMEN DEL CENSO================


,fuente,variable,tipo_python,n_faltantes,porcentaje_faltantes,valores_distintos
0,Lineal_3AP_1.dbf,COD_MOD,object,0,0.0,37354
1,Lineal_3AP_1.dbf,ANEXO,object,0,0.0,1
2,Lineal_3AP_1.dbf,CODLOCAL,object,0,0.0,37349
3,Lineal_3AP_1.dbf,NROCED,object,0,0.0,1
4,Lineal_3AP_1.dbf,CEN_EDU,object,0,0.0,33317
5,Lineal_3AP_1.dbf,DISTRITO,object,0,0.0,2899
6,Lineal_3AP_1.dbf,LOCALIDAD,object,0,0.0,20115
7,Lineal_3AP_1.dbf,CODGEO,object,0,0.0,1889
8,Lineal_3AP_1.dbf,NIV_MOD,object,0,0.0,1
9,Lineal_3AP_1.dbf,GES_DEP,object,0,0.0,10


### <font color=skyblue> 1.6 Limpieza de datos — Lineal_3AP_1 </font>

In [ ]:
faltantes_censo = pd.DataFrame({
    "n_faltantes": df_censo_final.isna().sum(),
    "porcentaje": (df_censo_final.isna().mean() * 100).round(2)
})

display(faltantes_censo)

,n_faltantes,porcentaje
COD_MOD,0,0.0
ANEXO,0,0.0
CODLOCAL,0,0.0
NROCED,0,0.0
CEN_EDU,0,0.0
DISTRITO,0,0.0
LOCALIDAD,0,0.0
CODGEO,0,0.0
NIV_MOD,0,0.0
GES_DEP,0,0.0


In [ ]:
print('--- P106A ---')
display(df_censo_final['P106A'].value_counts(dropna=False))

print('\n--- P107A_NO ---')
display(df_censo_final['P107A_NO'].value_counts(dropna=False))

print('\n--- P112A ---')
display(df_censo_final['P112A'].value_counts(dropna=False))

print('\n--- P113A ---')
display(df_censo_final['P113A'].value_counts(dropna=False))

print('\n--- P114A ---')
display(df_censo_final['P114A'].value_counts(dropna=False))

--- P106A ---


,count
P106A,
SI,22445
NO,14878
,31



--- P107A_NO ---


,count
P107A_NO,
,35821
1,867
4,270
2,209
3,187



--- P112A ---


,count
P112A,
NO,24991
SI,12333
,30



--- P113A ---


,count
P113A,
NO,17537
SI,16886
NS,2896
,35



--- P114A ---


,count
P114A,
SI,29648
NO,7706


In [ ]:
print('--- P116A (Sexo) ---')
display(df_censo_final['P116A'].value_counts(dropna=False))

print('\n--- P118A_S_1 ---')
display(df_censo_final['P118A_S_1'].value_counts(dropna=False))

print('\n--- P118A_S_3 ---')
display(df_censo_final['P118A_S_3'].value_counts(dropna=False))

print('\n--- P121A ---')
display(df_censo_final['P121A'].value_counts(dropna=False))

print('\n--- P122A ---')
display(df_censo_final['P122A'].value_counts(dropna=False))

--- P116A (Sexo) ---


,count
P116A,
SI,16482
NO,13158
,7714



--- P118A_S_1 ---


,count
P118A_S_1,
X,23152
,14202



--- P118A_S_3 ---


,count
P118A_S_3,
,24349
X,13005



--- P121A ---


,count
P121A,
8,16763
3,4952
9,4545
6,4064
5,2137
7,1670
1,1483
2,1247
4,491



--- P122A ---


,count
P122A,
SI,18917
NO,18406
,31


In [ ]:
print('--- P103B ---')
display(df_censo_final['P103B'].value_counts(dropna=False))

print('\n--- P105B ---')
display(df_censo_final['P105B'].value_counts(dropna=False))

print('\n--- P106B ---')
display(df_censo_final['P106B'].value_counts(dropna=False))

print('\n--- P109B_4 ---')
display(df_censo_final['P109B_4'].value_counts(dropna=False))

print('\n--- P109B_6 ---')
display(df_censo_final['P109B_6'].value_counts(dropna=False))

--- P103B ---


,count
P103B,
SI,31230
NO,6100
,24



--- P105B ---


,count
P105B,
SI,26023
,5828
NO,5503



--- P106B ---


,count
P106B,
SI,18389
,10680
NO,8285



--- P109B_4 ---


,count
P109B_4,
X,19499
,17855



--- P109B_6 ---


,count
P109B_6,
,24969
X,12385


In [ ]:
print('--- P110B_3 ---')
display(df_censo_final['P110B_3'].value_counts(dropna=False))

print('\n--- P114B_3 ---')
display(df_censo_final['P114B_3'].value_counts(dropna=False))

print('\n--- P114B_4 ---')
display(df_censo_final['P114B_4'].value_counts(dropna=False))

print('\n--- P114B_5 ---')
display(df_censo_final['P114B_5'].value_counts(dropna=False))

print('\n--- P114B_6 ---')
display(df_censo_final['P114B_6'].value_counts(dropna=False))

print('\n--- P114B_7 ---')
display(df_censo_final['P114B_7'].value_counts(dropna=False))

--- P110B_3 ---


,count
P110B_3,
X,22154
,15200



--- P114B_3 ---


,count
P114B_3,
,27349
X,10005



--- P114B_4 ---


,count
P114B_4,
,30676
X,6678



--- P114B_5 ---


,count
P114B_5,
,27794
X,9560



--- P114B_6 ---


,count
P114B_6,
,22940
X,14414



--- P114B_7 ---


,count
P114B_7,
,26339
X,11014
1,1


In [ ]:
print('--- P122B ---')
display(df_censo_final['P122B'].value_counts(dropna=False))

print('\n--- P123B ---')
display(df_censo_final['P123B'].value_counts(dropna=False))

print('\n--- P124B ---')
display(df_censo_final['P124B'].value_counts(dropna=False))

print('\n--- P124B_SI ---')
display(df_censo_final['P124B_SI'].value_counts(dropna=False))

print('\n--- P124B_NO ---')
display(df_censo_final['P124B_NO'].value_counts(dropna=False))

--- P122B ---


,count
P122B,
SI,31898
NO,3306
,2150



--- P123B ---


,count
P123B,
SI,32531
NO,4800
,23



--- P124B ---


,count
P124B,
SI,30709
NO,6611
,34



--- P124B_SI ---


,count
P124B_SI,
0,27474
,6591
1,1927
2,737
3,265
4,124
5,74
6,47
7,25



--- P124B_NO ---


,count
P124B_NO,
,30698
1,2897
4,2131
3,921
5,415
2,292


In [ ]:
print('--- P125B ---')
display(df_censo_final['P125B'].value_counts(dropna=False))

print('\n--- P126B_SI_1 ---')
display(df_censo_final['P126B_SI_1'].value_counts(dropna=False))

print('\n--- P127B ---')
display(df_censo_final['P127B'].value_counts(dropna=False))

print('\n--- P128B ---')
display(df_censo_final['P128B'].value_counts(dropna=False))

print('\n--- P129B ---')
display(df_censo_final['P129B'].value_counts(dropna=False))

--- P125B ---


,count
P125B,
SI,31045
NO,6269
,40



--- P126B_SI_1 ---


,count
P126B_SI_1,
,32721
1,1668
2,961
3,540
4,325
5,243
0,223
6,140
10,106



--- P127B ---


,count
P127B,
NO,20831
SI,16498
,25



--- P128B ---


,count
P128B,
SI,32517
NO,4812
,25



--- P129B ---


,count
P129B,
SI,33156
NO,4174
,24


In [ ]:
print("Filas duplicadas exactas en df_censo_final:", df_censo_final.duplicated().sum())

print("\nIDs repetidos (COD_MOD + ANEXO) en df_censo_final:")
display(df_censo_final[df_censo_final.duplicated(subset=["COD_MOD", "ANEXO"], keep=False)].sort_values(["COD_MOD", "ANEXO"]))

Filas duplicadas exactas en df_censo_final: 0

IDs repetidos (COD_MOD + ANEXO) en df_censo_final:


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO,NIV_MOD,GES_DEP,...,P122B,P123B,P124B,P124B_SI,P124B_NO,P125B,P126B_SI_1,P127B,P128B,P129B


### <font color=skyblue> 1.7 Conversión de tipos de datos — Lineal_3AP_1 </font>
Identificamos variables que deberían ser numéricas (IDs o conteos) y las convertimos. Usamos `errors='coerce'` para manejar valores no numéricos como NaN.

In [ ]:
# Variables identificadas para conversión numérica
df_censo_final['P126B_SI_1'] = pd.to_numeric(df_censo_final['P126B_SI_1'], errors='coerce')
print("Nueva verificación de tipos en df_censo_final:")
display(df_censo_final[['COD_MOD', 'ANEXO', 'CODLOCAL', 'CODGEO','P126B_SI_1']].dtypes)

Nueva verificación de tipos en df_censo_final:


,0
COD_MOD,object
ANEXO,object
CODLOCAL,object
CODGEO,object
P126B_SI_1,float64


### <font color=skyblue> 1.8 Estandarización de variables de texto (Sin bucles) — Lineal_3AP_1 </font>

Limpiamos espacios y normalizamos el formato de las columnas descriptivas y de respuesta usando operaciones vectorizadas.

In [ ]:
# Estandarización de nombres geográficos y categorías
df_censo_final['DISTRITO'] = df_censo_final['DISTRITO'].astype(str).str.strip().str.title()
df_censo_final['LOCALIDAD'] = df_censo_final['LOCALIDAD'].astype(str).str.strip().str.title()
df_censo_final['CEN_EDU'] = df_censo_final['CEN_EDU'].astype(str).str.strip().str.title()
df_censo_final['NIV_MOD'] = df_censo_final['NIV_MOD'].astype(str).str.strip().str.title()
df_censo_final['GES_DEP'] = df_censo_final['GES_DEP'].astype(str).str.strip().str.title()

# Estandarización de variables de respuesta SI/NO y Sexo
df_censo_final['P116A'] = df_censo_final['P116A'].astype(str).str.strip().str.upper()
df_censo_final['P129B'] = df_censo_final['P129B'].astype(str).str.strip().str.upper()
df_censo_final['P124B_SI'] = df_censo_final['P124B_SI'].astype(str).str.strip().str.upper()
df_censo_final['P124B_NO'] = df_censo_final['P124B_NO'].astype(str).str.strip().str.upper()

print("Estandarización finalizada para las columnas principales.")
display(df_censo_final[['DISTRITO', 'LOCALIDAD', 'P129B', 'P116A']].head())

Estandarización finalizada para las columnas principales.


,DISTRITO,LOCALIDAD,P129B,P116A
0,Cercado,,SI,
1,Alca,Anexo De Cahuana,SI,SI
2,Pillco Marca,Cayhuayna,SI,SI
3,Puente Piedra,Santa Paula,SI,NO
4,Yura,,SI,


In [ ]:
# Verificación de valores únicos sin bucles
print("Valores en P116A:")
display(df_censo_final['P116A'].value_counts())

print("\nValores en P129B:")
display(df_censo_final['P129B'].value_counts())

Valores en P116A:


,count
P116A,
SI,16482
NO,13158
,7714



Valores en P129B:


,count
P129B,
SI,33156
NO,4174
,24


In [ ]:
# Estandarización vectorizada de variables adicionales de respuesta
df_censo_final['P118A_S_1'] = df_censo_final['P118A_S_1'].astype(str).str.strip().str.upper()
df_censo_final['P118A_S_3'] = df_censo_final['P118A_S_3'].astype(str).str.strip().str.upper()
df_censo_final['P121A'] = df_censo_final['P121A'].astype(str).str.strip().str.upper()
df_censo_final['P122A'] = df_censo_final['P122A'].astype(str).str.strip().str.upper()
df_censo_final['P103B'] = df_censo_final['P103B'].astype(str).str.strip().str.upper()
df_censo_final['P105B'] = df_censo_final['P105B'].astype(str).str.strip().str.upper()
df_censo_final['P106B'] = df_censo_final['P106B'].astype(str).str.strip().str.upper()
df_censo_final['P122B'] = df_censo_final['P122B'].astype(str).str.strip().str.upper()
df_censo_final['P123B'] = df_censo_final['P123B'].astype(str).str.strip().str.upper()
df_censo_final['P125B'] = df_censo_final['P125B'].astype(str).str.strip().str.upper()
df_censo_final['P127B'] = df_censo_final['P127B'].astype(str).str.strip().str.upper()
df_censo_final['P128B'] = df_censo_final['P128B'].astype(str).str.strip().str.upper()

print("Estandarización de variables de respuesta completada sin el uso de bucles.")
display(df_censo_final[['P118A_S_1', 'P121A', 'P125B']].head())

Estandarización de variables de respuesta completada sin el uso de bucles.


,P118A_S_1,P121A,P125B
0,,6,SI
1,X,8,SI
2,X,9,SI
3,,6,SI
4,,1,SI


In [ ]:
# Estandarización vectorizada de variables adicionales
df_censo_final['P118A_S_1'] = df_censo_final['P118A_S_1'].astype(str).str.strip().str.upper()
df_censo_final['P118A_S_3'] = df_censo_final['P118A_S_3'].astype(str).str.strip().str.upper()
df_censo_final['P121A'] = df_censo_final['P121A'].astype(str).str.strip().str.upper()
df_censo_final['P122A'] = df_censo_final['P122A'].astype(str).str.strip().str.upper()
df_censo_final['P103B'] = df_censo_final['P103B'].astype(str).str.strip().str.upper()
df_censo_final['P105B'] = df_censo_final['P105B'].astype(str).str.strip().str.upper()
df_censo_final['P106B'] = df_censo_final['P106B'].astype(str).str.strip().str.upper()
df_censo_final['P122B'] = df_censo_final['P122B'].astype(str).str.strip().str.upper()
df_censo_final['P123B'] = df_censo_final['P123B'].astype(str).str.strip().str.upper()
df_censo_final['P125B'] = df_censo_final['P125B'].astype(str).str.strip().str.upper()
df_censo_final['P127B'] = df_censo_final['P127B'].astype(str).str.strip().str.upper()
df_censo_final['P128B'] = df_censo_final['P128B'].astype(str).str.strip().str.upper()

print("Estandarización de variables de respuesta completada.")
display(df_censo_final[['P118A_S_1', 'P121A', 'P125B']].head())

Estandarización de variables de respuesta completada.


,P118A_S_1,P121A,P125B
0,,6,SI
1,X,8,SI
2,X,9,SI
3,,6,SI
4,,1,SI


### <font color=skyblue> 1.9. Detección de valores fuera de rango — Lineal_3AP_1 </font>

Buscamos registros con valores inconsistentes o extremos en las variables numéricas y de identificación.

In [ ]:
# Detectar valores fuera de rango en variables numéricas
print("Registros con P126B_SI_1 fuera de rango (ej. < 0 o > 100):")
display(df_censo_final[(df_censo_final['P126B_SI_1'] < 0) | (df_censo_final['P126B_SI_1'] > 200)])



Registros con P126B_SI_1 fuera de rango (ej. < 0 o > 100):


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO,NIV_MOD,GES_DEP,...,P122B,P123B,P124B,P124B_SI,P124B_NO,P125B,P126B_SI_1,P127B,P128B,P129B
14671,0806521,0,495181,3AP,64935,Manantay,Pucallpa,250107,B0,A1,...,SI,SI,SI,0,,SI,230.0,NO,SI,SI
18528,1728054,0,806574,3AP,Futura Schools,Veintiseis De Octubre,Piura,200115,B0,B4,...,SI,SI,SI,1,,SI,387.0,SI,SI,SI
20486,0208348,0,139926,3AP,4011 La Taboada De Las Fresas,Callao,Callao,070101,B0,A1,...,SI,SI,SI,0,,SI,217.0,SI,SI,SI
22076,0606467,0,037466,3AP,88274,Santa,San Juan De Vinzos,021808,B0,A1,...,SI,SI,SI,0,,SI,217.0,SI,SI,SI
30956,1788686,0,864682,3AP,Educa,Villa María Del Triunfo,Vmt,150143,B0,B4,...,NO,NO,SI,0,,SI,217.0,SI,SI,SI
34915,0354423,0,412930,3AP,Salesiano Don Bosco,Castilla,Urbanizacion Miraflores,200104,B0,B2,...,SI,SI,SI,4,,SI,1000.0,NO,SI,SI


### <font color=skyblue> 1.10 Tratamiento de valores inválidos y faltantes (Sin bucles) — Lineal_3AP_1 </font>

Convertimos valores fuera de rango en NaN y aplicamos imputación simple para asegurar la completitud del dataset.

In [ ]:
# 1. Convertir valores numéricos inválidos a NaN
# Usamos el umbral de 1000 identificado en el paso anterior para P126B_SI_1
df_censo_final.loc[(df_censo_final['P126B_SI_1'] < 0) | (df_censo_final['P126B_SI_1'] > 1000), 'P126B_SI_1'] = np.nan

# 2. Imputación simple de valores numéricos
df_censo_final['P126B_SI_1'] = df_censo_final['P126B_SI_1'].fillna(df_censo_final['P126B_SI_1'].median())

# 3. Imputación de variables categóricas y de texto
# Reemplazamos nulos o strings vacíos (que quedaron tras el strip) por 'No especificado'
df_censo_final['DISTRITO'] = df_censo_final['DISTRITO'].replace(['', 'Nan', 'None'], np.nan).fillna('No especificado')
df_censo_final['LOCALIDAD'] = df_censo_final['LOCALIDAD'].replace(['', 'Nan', 'None'], np.nan).fillna('No especificado')
df_censo_final['P116A'] = df_censo_final['P116A'].replace(['', 'Nan', 'None'], np.nan).fillna('No especificado')
df_censo_final['P129B'] = df_censo_final['P129B'].replace(['', 'Nan', 'None'], np.nan).fillna('No especificado')
df_censo_final['NIV_MOD'] = df_censo_final['NIV_MOD'].replace(['', 'Nan', 'None'], np.nan).fillna('No especificado')
df_censo_final['GES_DEP'] = df_censo_final['GES_DEP'].replace(['', 'Nan', 'None'], np.nan).fillna('No especificado')

print("Tratamiento de faltantes completado.")
print("Faltantes restantes en columnas clave:")
display(df_censo_final[['P126B_SI_1', 'DISTRITO', 'LOCALIDAD', 'P116A', 'P129B']].isna().sum())

Tratamiento de faltantes completado.
Faltantes restantes en columnas clave:


,0
P126B_SI_1,0
DISTRITO,0
LOCALIDAD,0
P116A,0
P129B,0


### <font color=skyblue> 1.11 Gestión de Duplicados en df_censo_final — Lineal_3AP_1 </font>

Identificamos y eliminamos registros duplicados basados en la combinación de `COD_MOD` y `ANEXO`.

In [ ]:
# Identificar duplicados por la llave (COD_MOD, ANEXO)
duplicados_ids = df_censo_final.duplicated(subset=['COD_MOD', 'ANEXO'], keep=False).sum()
print(f"Total de IDs (COD_MOD + ANEXO) con duplicados: {duplicados_ids}")

# Eliminar duplicados manteniendo el primero
df_censo_final_limpio = df_censo_final.drop_duplicates(subset=['COD_MOD', 'ANEXO'], keep='first')

print(f"Dimensiones antes: {df_censo_final.shape}")
print(f"Dimensiones después de limpiar duplicados: {df_censo_final_limpio.shape}")

Total de IDs (COD_MOD + ANEXO) con duplicados: 0
Dimensiones antes: (37354, 41)
Dimensiones después de limpiar duplicados: (37354, 41)


In [ ]:
# Verificación final de duplicados en el nuevo dataset
duplicados_restantes = df_censo_final_limpio.duplicated(subset=['COD_MOD', 'ANEXO']).sum()
print(f"Duplicados restantes en df_censo_final_limpio: {duplicados_restantes}")

Duplicados restantes en df_censo_final_limpio: 0


In [ ]:
df_censo_final_limpio.head(5)

,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO,NIV_MOD,GES_DEP,...,P122B,P123B,P124B,P124B_SI,P124B_NO,P125B,P126B_SI_1,P127B,P128B,P129B
0,1720556,0,852373,3AP,Sir Alexander Fleming,Cercado,No especificado,040101,B0,B4,...,SI,SI,NO,,2,SI,2.0,SI,SI,SI
1,0313890,0,073181,3AP,40513 San Miguel Arcangel,Alca,Anexo De Cahuana,040802,B0,A1,...,SI,SI,SI,0,,SI,2.0,NO,SI,SI
2,1495092,0,605028,3AP,33421,Pillco Marca,Cayhuayna,100111,B0,A1,...,SI,SI,SI,0,,SI,1.0,SI,SI,SI
3,1579861,0,651751,3AP,Santander De Puente Piedra,Puente Piedra,Santa Paula,150125,B0,B4,...,NO,SI,SI,0,,SI,2.0,NO,NO,SI
4,1529767,0,553166,3AP,Colegio Superior De Ciencias Newton Vianney,Yura,No especificado,040128,B0,B4,...,NO,SI,SI,0,,SI,2.0,NO,NO,SI


### <font color=skyblue> 1.12 Justificación de Decisiones de Limpieza y Preprocesamiento — Lineal_3AP_1 </font>

En esta etapa se transformó el dataset crudo `Lineal_3AP_1.dbf` en el dataset procesado `df_censo_final_limpio` siguiendo estas acciones:

1.  **Correcciones y Estandarización:**
    *   **Nombres de Variables:** Se convirtieron todos los nombres de las columnas a mayúsculas para asegurar consistencia.
    *   **Normalización de Texto:** Se eliminaron espacios en blanco y se aplicó formato `Title Case` a nombres geográficos (`DISTRITO`, `LOCALIDAD`) y `UPPERCASE` a variables de respuesta (`SI`, `NO`, `SEXO`) mediante métodos vectorizados para evitar errores de digitación.
    *   **Conversión de Tipos:** Los identificadores (`COD_MOD`, `ANEXO`, `CODLOCAL`, `CODGEO`) y variables de conteo (`P126B_SI_1`) se convirtieron a formatos numéricos (`int64`/`float64`) para habilitar cálculos estadísticos.

2.  **Imputación de Datos:**
    *   **Valores Numéricos:** Se detectaron 9 valores atípicos en `P126B_SI_1` (Variable: En el año 2024, ¿se hizo uso del libro de registro de incidencias?) (superiores a 200, llegando hasta 1000). Estos se trataron como `NaN` y se imputaron utilizando la **mediana** de la distribución para no sesgar el análisis con valores extremos.
    *   **Valores Categóricos:** Los registros nulos o vacíos en variables críticas se reemplazaron con la etiqueta **'No especificado'**, manteniendo la integridad de las filas sin perder información contextual.

3.  **Eliminación y Mantenimiento:**
    *   **Duplicados:** Se validó la llave primaria compuesta (`COD_MOD`, `ANEXO`). Tras la verificación inicial, se confirmó que el dataset no presentaba duplicados exactos ni colisiones de ID, manteniendo los 37,354 registros originales.
    *   **Filtrado:** Se mantuvieron únicamente las 41 variables requeridas para el análisis del laboratorio, eliminando las columnas sobrantes del archivo `.dbf` original para optimizar el uso de memoria.

## <font color=yellow> 2. Tratamiento de datos — denuncias_escolares </font>

Aplicamos el mismo chequeo sistemático usado para F1 y F2, esta vez sobre `df2` (denuncias_escolares)

### <font color=skyblue> 2.1 Importación de datos — denuncias_escolares </font>

In [ ]:
import pandas as pd

# Cargar la base de datos saltando las filas de encabezado informativo (metadata)
ruta_denuncias = 'denuncias_escolares.xlsx'
df2 = pd.read_excel(ruta_denuncias, skiprows=5)

print(f"Base de datos 'denuncias_escolares' cargada con {df2.shape[0]} filas y {df2.shape[1]} columnas.")
display(df2.head())

Base de datos 'denuncias_escolares' cargada con 48273 filas y 8 columnas.


,FECHA_REPORTE,DRE,UGEL,NIVEL_EDUCATIVO,TIPO_REPORTE,TIPO_VIOLENCIA,SUBTIPO_VIOLENCIA,TIPO_ESTADO_REPORTE
0,2026-07-31,DRE Tacna,UGEL Tacna,Primaria,Personal IE a Escolares,Sexual,Hostigamiento sexual,Atención en proceso
1,2026-07-31,DRE Tacna,UGEL Tacna,Primaria,Personal IE a Escolares,Sexual,"Tocamientos, actos de connotación sexual o act...",Atención en proceso
2,2026-07-31,DRE Arequipa,UGEL Arequipa Sur,Secundaria,Entre Escolares,Psicológica,Verbal,Pendiente de atención por la IE
3,2026-07-31,DRE Ica,UGEL Ica,Secundaria,Entre Escolares,Física,Sin lesiones,Pendiente de atención por la IE
4,2026-07-31,DRE Cusco,UGEL Anta,Secundaria,Entre Escolares,Psicológica,Con uso de armas,Pendiente de atención por la IE


### <font color=skyblue> 2.2 Chequeo de faltantes en datos — denuncias_escolares </font>


In [ ]:
vacio_denuncias = df2.isna() | (df2.astype(str).apply(lambda s: s.str.strip()) == "")

faltantes_denuncias = pd.DataFrame({
    "n_faltantes": vacio_denuncias.sum(),
    "porcentaje": (vacio_denuncias.mean() * 100).round(2)
})

print("Faltantes en denuncias_escolares:")
display(faltantes_denuncias)

Faltantes en denuncias_escolares:


,n_faltantes,porcentaje
FECHA_REPORTE,0,0.0
DRE,0,0.0
UGEL,0,0.0
NIVEL_EDUCATIVO,0,0.0
TIPO_REPORTE,0,0.0
TIPO_VIOLENCIA,0,0.0
SUBTIPO_VIOLENCIA,0,0.0
TIPO_ESTADO_REPORTE,0,0.0


### <font color=skyblue> 2.3 Chequeo de duplicados en datos — denuncias_escolares </font>


In [ ]:
# denuncias_escolares no trae un ID de caso: no podemos revisar duplicados por llave,
# solo duplicados exactos (misma fecha + misma combinación de categorías).
print("Filas duplicadas exactas en denuncias_escolares:", df2.duplicated().sum())
print("Filas dentro de algún grupo duplicado:", df2.duplicated(keep=False).sum())

print("\nEjemplo de un grupo de filas duplicadas:")
display(df2[df2.duplicated(keep=False)].sort_values(list(df2.columns)).head(10))

Filas duplicadas exactas en denuncias_escolares: 7282
Filas dentro de algún grupo duplicado: 12217

Ejemplo de un grupo de filas duplicadas:


,FECHA_REPORTE,DRE,UGEL,NIVEL_EDUCATIVO,TIPO_REPORTE,TIPO_VIOLENCIA,SUBTIPO_VIOLENCIA,TIPO_ESTADO_REPORTE
48264,2024-01-02,DRE Tacna,UGEL Tacna,Secundaria,Personal IE a Escolares,Psicológica,Trato humillante,Atención finalizada
48265,2024-01-02,DRE Tacna,UGEL Tacna,Secundaria,Personal IE a Escolares,Psicológica,Trato humillante,Atención finalizada
48266,2024-01-02,DRE Tacna,UGEL Tacna,Secundaria,Personal IE a Escolares,Psicológica,Trato humillante,Atención finalizada
48267,2024-01-02,DRE Tacna,UGEL Tacna,Secundaria,Personal IE a Escolares,Psicológica,Trato humillante,Atención finalizada
48240,2024-01-04,DRE Arequipa,UGEL Camaná,Secundaria,Personal IE a Escolares,Sexual,"Tocamientos, actos de connotación sexual o act...",Atención finalizada
48241,2024-01-04,DRE Arequipa,UGEL Camaná,Secundaria,Personal IE a Escolares,Sexual,"Tocamientos, actos de connotación sexual o act...",Atención finalizada
48242,2024-01-04,DRE Cusco,UGEL Pichari-Kimbiri-Villa Virgen,Secundaria,Personal IE a Escolares,Sexual,Violencia con fines sexuales a través de medio...,Atención finalizada
48243,2024-01-04,DRE Cusco,UGEL Pichari-Kimbiri-Villa Virgen,Secundaria,Personal IE a Escolares,Sexual,Violencia con fines sexuales a través de medio...,Atención finalizada
48244,2024-01-04,DRE Cusco,UGEL Pichari-Kimbiri-Villa Virgen,Secundaria,Personal IE a Escolares,Sexual,Violencia con fines sexuales a través de medio...,Atención finalizada
48209,2024-01-11,DRE Cusco,UGEL Canchis,Inicial - Cuna-Jardín,Personal IE a Escolares,Psicológica,Verbal,Atención finalizada


### <font color=skyblue> 2.4 Tipos de datos — denuncias_escolares </font>

In [ ]:
print("Tipos detectados en denuncias_escolares:")
display(df2.dtypes)
# FECHA_REPORTE ya se reconoce como datetime64 al leer el Excel; el resto son
# variables categóricas (texto), que es el tipo correcto: no hay columnas numéricas
# continuas en esta fuente, así que no aplica conversión con pd.to_numeric.

Tipos detectados en denuncias_escolares:


,0
FECHA_REPORTE,datetime64[ns]
DRE,object
UGEL,object
NIVEL_EDUCATIVO,object
TIPO_REPORTE,object
TIPO_VIOLENCIA,object
SUBTIPO_VIOLENCIA,object
TIPO_ESTADO_REPORTE,object


### <font color=skyblue> 2.5 Valores fuera de rango — denuncias_escolares </font>

In [ ]:
print("Rango de fechas de reporte:", df2["FECHA_REPORTE"].min(), "a", df2["FECHA_REPORTE"].max())
# No hay fechas futuras respecto al corte de la fuente ni fechas absurdamente
# antiguas: el rango es razonable y no requiere corrección.

Rango de fechas de reporte: 2024-01-02 00:00:00 a 2026-07-31 00:00:00


### <font color=skyblue> 2.6 Posibles outliers — denuncias_escolares </font>

In [ ]:
# Sin variables numéricas, los "outliers" aquí se entienden como categorías de
# muy baja frecuencia, que conviene revisar antes de agrupar o modelar.
print("Niveles educativos poco frecuentes:")
display(df2["NIVEL_EDUCATIVO"].value_counts(dropna=False).tail(5))

print("\nSubtipos de violencia poco frecuentes:")
display(df2["SUBTIPO_VIOLENCIA"].value_counts(dropna=False).tail(5))

Niveles educativos poco frecuentes:


,count
NIVEL_EDUCATIVO,
Inicial - Cuna-Jardín,302
Básica Alternativa - Avanzado,127
Básica Especial - Primaria,86
Básica Especial - Inicial,16
Básica Alternativa - Inicial e Intermedio,9



Subtipos de violencia poco frecuentes:


,count
SUBTIPO_VIOLENCIA,
Violencia con fines sexuales a través de medios tecnológicos,425
Acoso sexual,335
Acoso sexual a través de medios tecnológicos,297
Aislamiento,275
Con uso de armas,70


### <font color=skyblue> 2.7 Categorias no uniformes — denuncias_escolares </font>

In [ ]:
# Verificamos espacios extremos y duplicados por mayúsculas/minúsculas en cada
# variable categórica de texto.
for col in ["DRE", "UGEL", "NIVEL_EDUCATIVO", "TIPO_REPORTE", "TIPO_VIOLENCIA",
            "SUBTIPO_VIOLENCIA", "TIPO_ESTADO_REPORTE"]:
    con_espacios = (df2[col] != df2[col].str.strip()).sum()
    valores_unicos = df2[col].unique()
    posibles_dup = pd.Series(valores_unicos).str.lower().str.strip().nunique() != len(valores_unicos)
    print(f"{col}: filas con espacios extra = {con_espacios} | posibles duplicados por may/min = {posibles_dup}")
# Resultado: no se detectan inconsistencias de formato en esta fuente.

DRE: filas con espacios extra = 0 | posibles duplicados por may/min = False
UGEL: filas con espacios extra = 0 | posibles duplicados por may/min = False
NIVEL_EDUCATIVO: filas con espacios extra = 0 | posibles duplicados por may/min = False
TIPO_REPORTE: filas con espacios extra = 0 | posibles duplicados por may/min = False
TIPO_VIOLENCIA: filas con espacios extra = 0 | posibles duplicados por may/min = False
SUBTIPO_VIOLENCIA: filas con espacios extra = 0 | posibles duplicados por may/min = False
TIPO_ESTADO_REPORTE: filas con espacios extra = 0 | posibles duplicados por may/min = False


### <font color=skyblue>2.8 Decisiones de limpieza justificadas — denuncias_escolares </font>

Registramos qué se corrigió, qué se dejó documentado sin modificar y por qué, antes de aplicar los cambios.

In [ ]:
decisiones_denuncias = pd.DataFrame([
    ["UGEL", "En 1,328 filas UGEL repite el nombre de la DRE ('DRE Callao')",
     "Mantener sin modificar, documentado como válido",
     "Callao es la única región sin sub-UGELs (una sola UGEL cubre toda la DRE); no es un error de captura."],
    ["TIPO_VIOLENCIA / SUBTIPO_VIOLENCIA", "19 casos de 'Con uso de armas' bajo violencia Psicológica",
     "Mantener y marcar para revisión manual, sin reclasificar automáticamente",
     "No existe una regla de negocio clara para decidir si es error de tipeo o una clasificación válida del reportante."],
    ["Fila completa (todas las variables)", "7,282 filas duplicadas exactas; la fuente no trae un ID de caso",
     "Eliminar duplicados exactos, conservando el primer registro (keep='first')",
     "Sin ID de caso no se puede distinguir una denuncia repetida de dos denuncias distintas con iguales atributos; se opta por el criterio conservador de no inflar el conteo de casos."],
    ["DRE, UGEL, NIVEL_EDUCATIVO, TIPO_REPORTE, TIPO_VIOLENCIA, SUBTIPO_VIOLENCIA, TIPO_ESTADO_REPORTE",
     "Vienen como texto (object/string)",
     "Convertir a dtype category",
     "Son variables nominales de baja cardinalidad; category deja explícito su rol y hace más eficiente el análisis posterior."],
], columns=["variable", "problema", "decisión", "justificación"])

display(decisiones_denuncias)

,variable,problema,decisión,justificación
0,UGEL,"En 1,328 filas UGEL repite el nombre de la DRE...","Mantener sin modificar, documentado como válido",Callao es la única región sin sub-UGELs (una s...
1,TIPO_VIOLENCIA / SUBTIPO_VIOLENCIA,19 casos de 'Con uso de armas' bajo violencia ...,"Mantener y marcar para revisión manual, sin re...",No existe una regla de negocio clara para deci...
2,Fila completa (todas las variables),"7,282 filas duplicadas exactas; la fuente no t...","Eliminar duplicados exactos, conservando el pr...",Sin ID de caso no se puede distinguir una denu...
3,"DRE, UGEL, NIVEL_EDUCATIVO, TIPO_REPORTE, TIPO...",Vienen como texto (object/string),Convertir a dtype category,Son variables nominales de baja cardinalidad; ...


### <font color=skyblue> 2.9 Aplicar las decisiones — denuncias_escolares </font>

In [ ]:
df2.dtypes

,0
FECHA_REPORTE,datetime64[ns]
DRE,object
UGEL,object
NIVEL_EDUCATIVO,object
TIPO_REPORTE,object
TIPO_VIOLENCIA,object
SUBTIPO_VIOLENCIA,object
TIPO_ESTADO_REPORTE,object


In [ ]:
# Partimos de df2 (denuncias_escolares) sin tocar df ni df1

# 1) Eliminar duplicados exactos, conservando el primer registro
df2_limpio = df2.drop_duplicates(keep="first").copy()
print("denuncias_escolares antes y después de quitar duplicados:", df2.shape, df2_limpio.shape)

print("\nTipos finales:")
display(df2_limpio.dtypes)

print("\nFaltantes finales:")
display(df2_limpio.isna().sum())

denuncias_escolares antes y después de quitar duplicados: (48273, 8) (40991, 8)

Tipos finales:


,0
FECHA_REPORTE,datetime64[ns]
DRE,object
UGEL,object
NIVEL_EDUCATIVO,object
TIPO_REPORTE,object
TIPO_VIOLENCIA,object
SUBTIPO_VIOLENCIA,object
TIPO_ESTADO_REPORTE,object



Faltantes finales:


,0
FECHA_REPORTE,0
DRE,0
UGEL,0
NIVEL_EDUCATIVO,0
TIPO_REPORTE,0
TIPO_VIOLENCIA,0
SUBTIPO_VIOLENCIA,0
TIPO_ESTADO_REPORTE,0


In [ ]:
df2.head(5)

,FECHA_REPORTE,DRE,UGEL,NIVEL_EDUCATIVO,TIPO_REPORTE,TIPO_VIOLENCIA,SUBTIPO_VIOLENCIA,TIPO_ESTADO_REPORTE
0,2026-07-31,DRE Tacna,UGEL Tacna,Primaria,Personal IE a Escolares,Sexual,Hostigamiento sexual,Atención en proceso
1,2026-07-31,DRE Tacna,UGEL Tacna,Primaria,Personal IE a Escolares,Sexual,"Tocamientos, actos de connotación sexual o act...",Atención en proceso
2,2026-07-31,DRE Arequipa,UGEL Arequipa Sur,Secundaria,Entre Escolares,Psicológica,Verbal,Pendiente de atención por la IE
3,2026-07-31,DRE Ica,UGEL Ica,Secundaria,Entre Escolares,Física,Sin lesiones,Pendiente de atención por la IE
4,2026-07-31,DRE Cusco,UGEL Anta,Secundaria,Entre Escolares,Psicológica,Con uso de armas,Pendiente de atención por la IE


In [ ]:
df2_limpio['UGEL_norm'] = df2_limpio['UGEL'].str.strip().str.upper()

# Filtrar y contar denuncias por tipo de violencia
df_fisica = df2_limpio[df2_limpio['TIPO_VIOLENCIA'] == 'Física'].groupby('UGEL_norm').size().reset_index(name='nro_denuncias_tipo_fisica')
df_psicologica = df2_limpio[df2_limpio['TIPO_VIOLENCIA'] == 'Psicológica'].groupby('UGEL_norm').size().reset_index(name='nro_denuncias_tipo_psicologica')
df_sexual = df2_limpio[df2_limpio['TIPO_VIOLENCIA'] == 'Sexual'].groupby('UGEL_norm').size().reset_index(name='nro_denuncias_tipo_sexual')

# Filtrar y contar denuncias por tipo de reporte
df_entreescolares = df2_limpio[df2_limpio['TIPO_REPORTE'] == 'Entre Escolares'].groupby('UGEL_norm').size().reset_index(name='nro_denuncias_tipo_reporte_entreescolares')
df_personalescolar = df2_limpio[df2_limpio['TIPO_REPORTE'] == 'Personal IE a Escolares'].groupby('UGEL_norm').size().reset_index(name='nro_denuncias_tipo_reporte_personalescolar')

# Unir los DataFrames
df_violencia_por_ugel = df_fisica.merge(df_psicologica, on='UGEL_norm', how='outer')\
                                 .merge(df_sexual, on='UGEL_norm', how='outer')\
                                 .merge(df_entreescolares, on='UGEL_norm', how='outer')\
                                 .merge(df_personalescolar, on='UGEL_norm', how='outer')

# Rellenar los valores NaN (donde no hay denuncias de un tipo específico o de reporte) con 0
df_violencia_por_ugel = df_violencia_por_ugel.fillna(0)

# Mostrar el resultado
display(df_violencia_por_ugel.head())

,UGEL_norm,nro_denuncias_tipo_fisica,nro_denuncias_tipo_psicologica,nro_denuncias_tipo_sexual,nro_denuncias_tipo_reporte_entreescolares,nro_denuncias_tipo_reporte_personalescolar
0,DRE CALLAO,485.0,510.0,160.0,806.0,349
1,UGEL 01 - EL PORVENIR,55.0,53.0,21.0,84.0,45
2,UGEL 01 SAN JUAN DE MIRAFLORES,759.0,908.0,311.0,1371.0,607
3,UGEL 02 - LA ESPERANZA,69.0,55.0,24.0,89.0,59
4,UGEL 02 RÍMAC,771.0,771.0,316.0,1306.0,552


## <font color=yellow> 3. Tratamiento de datos — Padron_web </font>

### <font color=skyblue> 3.1 Importación de datos — Padron_web </font>

In [ ]:
from dbfread import DBF

ruta1 = "Padron_web.dbf"

tabla_padron = DBF(ruta1, encoding="cp850", char_decode_errors="replace")

df_padron = pd.DataFrame(iter(tabla_padron))

display(df_padron.head(5))

,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT
0,24953981,0415547,0,016100,123,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,,11,Mañana,,,,1,Activo,,05-08-2026
1,25273230,0415638,0,015172,122,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,,11,Mañana,,,,1,Activo,02-05-1942,05-08-2026
2,20414731,0415646,0,015186,233,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,,13,Mañana-Tarde,,,,1,Activo,04-08-1974,05-08-2026
3,26281962,0415877,0,016751,COLEGIO PARROQUIAL NUESTRA SEÑORA DEL SAGRADO ...,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,,13,Mañana-Tarde,20146930418,C.E.P. NUESTRA SEÑORA DEL SAGRADO CORAZON DE J...,,1,Activo,07-05-1962,05-08-2026
4,25111788,0567206,0,016119,268,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,,11,Mañana,,,,1,Activo,24-04-1981,05-08-2026


### <font color=skyblue> 3.2 Chequeo de duplicados — Padron_web </font>


In [ ]:
#Padron_web
print("Filas duplicadas exactas en Padron_web:", df_padron.duplicated().sum())

CLAVE_PADRON = ["COD_MOD", "ANEXO"]

print("\nIDs repetidos en Padron_web:")
display(df_padron[df_padron.duplicated(subset=CLAVE_PADRON, keep=False)].sort_values(CLAVE_PADRON))


Filas duplicadas exactas en Padron_web: 0

IDs repetidos en Padron_web:


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,D_TIPOPROG,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT


### <font color=skyblue> 3.3 Conversión de tipos de datos — Padron_web </font> <font color=green>

In [ ]:
# df = Padron_web.dbf
padron = df_padron.copy()


padron["NLAT_IE"]  = pd.to_numeric(padron["NLAT_IE"], errors="coerce")
padron["NLONG_IE"] = pd.to_numeric(padron["NLONG_IE"], errors="coerce")


padron["FECHAREG"]  = pd.to_datetime(padron["FECHAREG"], errors="coerce", dayfirst=True)
padron["FECHA_ACT"] = pd.to_datetime(padron["FECHA_ACT"], errors="coerce", dayfirst=True)


padron["COD_MOD"]  = padron["COD_MOD"].astype(str).str.strip().str.zfill(7)
padron["ANEXO"]    = padron["ANEXO"].astype(str).str.strip()
padron["CODLOCAL"] = padron["CODLOCAL"].astype(str).str.strip().str.zfill(6)
padron["CODGEO"]   = padron["CODGEO"].astype(str).str.strip().str.zfill(6)
padron["CODOOII"]  = padron["CODOOII"].astype(str).str.strip().str.zfill(6)
padron["CODINST"]  = padron["CODINST"].astype(str).str.strip()


sin_coord = (padron["NLAT_IE"] == 0) & (padron["NLONG_IE"] == 0)
padron.loc[sin_coord, ["NLAT_IE", "NLONG_IE"]] = pd.NA


padron["TELEFONO"] = padron["TELEFONO"].replace("", pd.NA)
padron["EMAIL"]    = padron["EMAIL"].replace("", pd.NA)
padron["NRORUC"]   = padron["NRORUC"].replace("", pd.NA)
padron["PROMOTOR"] = padron["PROMOTOR"].replace("", pd.NA)

print("Tipos después de convertir:")
display(padron.dtypes)



Tipos después de convertir:


,0
CODINST,object
COD_MOD,object
ANEXO,object
CODLOCAL,object
CEN_EDU,object
NIV_MOD,object
D_NIV_MOD,object
D_FORMA,object
TIPSSEXO,object
D_TIPSSEXO,object


### <font color=skyblue> 3.4 Estandarización de nombres y categorías— Padron_web </font>

In [ ]:
#Padron_web
# Quitar espacios y unificar mayúsculas
padron["D_NIV_MOD"]  = padron["D_NIV_MOD"].astype("string").str.strip()
padron["D_FORMA"]    = padron["D_FORMA"].astype("string").str.strip()
padron["D_TIPSSEXO"] = padron["D_TIPSSEXO"].astype("string").str.strip()
padron["D_GESTION"]  = padron["D_GESTION"].astype("string").str.strip()
padron["D_GES_DEP"]  = padron["D_GES_DEP"].astype("string").str.strip()
padron["DAREACENSO"] = padron["DAREACENSO"].astype("string").str.strip()
padron["D_ESTADO"]   = padron["D_ESTADO"].astype("string").str.strip()
padron["D_COD_TUR"]  = padron["D_COD_TUR"].astype("string").str.strip()
padron["D_TIPOPROG"] = padron["D_TIPOPROG"].astype("string").str.strip()

padron["D_DPTO"]    = padron["D_DPTO"].astype("string").str.strip().str.upper()
padron["D_PROV"]    = padron["D_PROV"].astype("string").str.strip().str.upper()
padron["D_DIST"]    = padron["D_DIST"].astype("string").str.strip().str.upper()
padron["D_REGION"]  = padron["D_REGION"].astype("string").str.strip().str.upper()
padron["D_DREUGEL"] = padron["D_DREUGEL"].astype("string").str.strip().str.upper()
padron["EMAIL"]     = padron["EMAIL"].astype("string").str.strip().str.lower()

# Quitar tildes en las columnas
padron["D_REGION"] = (padron["D_REGION"].str.normalize("NFKD")
                      .str.encode("ascii", errors="ignore").str.decode("utf-8").astype("string"))
padron["D_DREUGEL"] = (padron["D_DREUGEL"].str.normalize("NFKD")
                       .str.encode("ascii", errors="ignore").str.decode("utf-8").astype("string"))

# Mapear categorías
padron["D_REGION"] = padron["D_REGION"].replace({
    "GRE AREQUIPA": "DRE AREQUIPA",
    "GRE LA LIBERTAD": "DRE LA LIBERTAD",
    "GRE LAMBAYEQUE": "DRE LAMBAYEQUE"
})

padron["D_TIPOPROG"] = padron["D_TIPOPROG"].replace({
    "": pd.NA,
    "\x00": pd.NA
})

padron["EMAIL"] = padron["EMAIL"].replace({
    "": pd.NA
})

padron["activo"] = padron["D_ESTADO"].map({
    "Activo": True,
    "Inactivo": False
}).astype("boolean")

print("Categorías estandarizadas:")
display(padron["D_GESTION"].value_counts(dropna=False))
display(padron["D_GES_DEP"].value_counts(dropna=False))
display(padron["DAREACENSO"].value_counts(dropna=False))
display(padron["D_REGION"].value_counts(dropna=False))
display(padron["D_TIPOPROG"].value_counts(dropna=False))
display(padron["activo"].value_counts(dropna=False))

Categorías estandarizadas:


,count
D_GESTION,
Pública de gestión directa,130758
Privada,48571
Pública de gestión privada,1463


,count
D_GES_DEP,
Sector Educación,126405
Particular,45380
Municipalidad,3832
Comunidad,1475
Convenio con Sector Educación,1463
Comunidad o asociación religiosa,949
Otro sector público (FF.AA.),521
Asociación civil / Inst.Benéfica,391
Empresa (Fiscalizado),262


,count
DAREACENSO,
Rural,92410
Urbana,88382


,count
D_REGION,
DRE LIMA METROPOLITANA,29917
DRE CAJAMARCA,13631
DRE PUNO,10524
DRE PIURA,10386
DRE CUSCO,9627
DRE LA LIBERTAD,9391
DRE JUNIN,9113
DRE AREQUIPA,8453
DRE ANCASH,8242


,count
D_TIPOPROG,
<NA>,124970
Ciclo II - Entorno Comunitario,18539
PRONOEI,16888
Ciclo I - Entorno Comunitario,9614
SET,2923
Ciclo I - Entorno Familiar,2514
PIETBAF,1674
Ciclo I - SET,1671
Ciclo II - Entorno Familiar,857


,count
activo,
True,115453
False,65339


### <font color=skyblue> 3.5 Valores fuera de rango — Padron_web </font>

In [ ]:
#Padron_web
HOY = pd.Timestamp("today").normalize()

print("Coordenadas fuera del territorio peruano:")
display(padron[(padron["NLAT_IE"] < -18.4) | (padron["NLAT_IE"] > 0) |
               (padron["NLONG_IE"] < -81.4) | (padron["NLONG_IE"] > -68.6)])

print("IE sin coordenadas (venían como 0,0):")
display(padron[padron["NLAT_IE"].isna() | padron["NLONG_IE"].isna()])

print("Fechas de creación en el futuro:")
display(padron[padron["FECHAREG"] > HOY])

print("Fechas de creación anteriores a 1900 (no necesariamente error):")
display(padron[padron["FECHAREG"] < "1900-01-01"])

print("Registros actualizados antes de haber sido creados:")
display(padron[padron["FECHA_ACT"] < padron["FECHAREG"]])

print("RUC con formato inválido:")
display(padron[padron["NRORUC"].notna() &
               ~padron["NRORUC"].str.match(r"^(10|15|16|17|20)\d{9}$").fillna(False)])

print("Correos con formato inválido:")
display(padron[padron["EMAIL"].notna() &
               ~padron["EMAIL"].str.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$").fillna(False)])

Coordenadas fuera del territorio peruano:


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo


IE sin coordenadas (venían como 0,0):


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo
297,,0416271,0,078198,87501,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,1959-03-10,2026-08-05,False
462,,0818500,0,017944,86853,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,1992-11-25,2026-08-05,False
706,,0385666,0,078234,86252,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,NaT,2026-08-05,False
711,,0681320,0,078239,86949,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,NaT,2026-08-05,False
712,,0385823,0,078240,86264-84,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,NaT,2026-08-05,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
166431,,1767888,0,838495,LOS OLIVOS,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,2018-07-24,2026-08-05,False
167833,,3960335,0,000000,LOS NIÑOS DE COSCORE,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,2019-03-04,2026-08-05,False
168942,,3966754,0,396675,AMIGUITOS DEL SABER,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,2019-05-13,2026-08-05,False
171177,,3976688,0,397668,GOTITAS DE AMOR,A5,Inicial - Programa no escolarizado,No escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,2020-02-25,2026-08-05,False


Fechas de creación en el futuro:


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo
10518,23480921,0271601,0,503483,64052 SANTA ROSA DE LIMA,B0,Primaria,Escolarizada,3,Mixto,...,13,Mañana-Tarde,<NA>,,<NA>,1,Activo,2055-09-08,2026-08-05,True


Fechas de creación anteriores a 1900 (no necesariamente error):


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo
25,26281709,0411728,0,015370,86019 LA LIBERTAD,B0,Primaria,Escolarizada,3,Mixto,...,13,Mañana-Tarde,<NA>,,<NA>,1,Activo,1828-01-30,2026-08-05,True
113,26281709,0577486,0,015370,86019 LA LIBERTAD,F0,Secundaria,Escolarizada,3,Mixto,...,13,Mañana-Tarde,<NA>,,<NA>,1,Activo,1828-01-30,2026-08-05,True
202,20419458,0412726,0,017449,86065,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1866-04-28,2026-08-05,True
1467,24954145,0587394,0,026627,DOS DE MAYO,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1869-01-28,2026-08-05,True
2151,20935731,0575282,0,039578,SAN PEDRO,F0,Secundaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1883-12-02,2026-08-05,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
80897,26956412,0429779,0,177409,37002 RICARDO FERNANDEZ,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1882-03-18,2026-08-05,True
81011,,0515064,0,827717,34027,L0,Técnico Productiva,Escolarizada,3,Mixto,...,14,Noche,<NA>,,<NA>,1,Activo,1893-07-01,2026-08-05,True
86705,,0923714,0,495893,65098,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,1898-08-04,2026-08-05,False
86708,,1120690,0,088868,457,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,1898-08-07,2026-08-05,False


Registros actualizados antes de haber sido creados:


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo
10518,23480921,0271601,0,503483,64052 SANTA ROSA DE LIMA,B0,Primaria,Escolarizada,3,Mixto,...,13,Mañana-Tarde,<NA>,,<NA>,1,Activo,2055-09-08,2026-08-05,True
180779,22378163,0873950,1,920211,NUEVO HORIZONTE,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180780,22378163,1476399,1,920211,NUEVO HORIZONTE,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180781,,1801414,1,868208,SAN FRANCISCO DE SALES,T0,Instituto Superior Tecnológico,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180782,27019935,3063682,0,920230,825,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180783,27020097,3063690,0,920249,35030 HELEN ADAMS KELLER,E2,Básica Especial-Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180784,27020178,3063708,0,920254,824,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180785,27020259,3063716,0,920268,LITTLE TALENTS SCHOOL,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,20609620456,LITTLE TALENTS SCHOOL S.A.C.,<NA>,1,Activo,2026-08-07,2026-08-05,True
180786,27020331,3063724,0,920273,55022 SAN PEDRO - SAN PABLO,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True
180787,21756441,3063732,0,207406,64569,F0,Secundaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,2026-08-07,2026-08-05,True


RUC con formato inválido:


/tmp/ipykernel_3695/4095401966.py:22: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  ~padron["NRORUC"].str.match(r"^(10|15|16|17|20)\d{9}$").fillna(False)])


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo
69664,26583837,0305748,0,295436,VIRGEN DE FATIMA DE BARRANCO,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,26162873262,MISIONERAS SEGLARES DE JESUS SACRAMENTADO Y DE...,<NA>,1,Activo,1983-10-25,2026-08-05,True
69719,26583837,0468686,0,295436,VIRGEN DE FATIMA DE BARRANCO,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,26162873262,MISIONERAS SEGLARES DE JESUS SACRAMENTADO Y DE...,<NA>,1,Activo,1968-05-02,2026-08-05,True
69749,26583837,0643239,0,295436,VIRGEN DE FATIMA DE BARRANCO,F0,Secundaria,Escolarizada,3,Mixto,...,11,Mañana,26162873262,MISIONERAS SEGLARES DE JESUS SACRAMENTADO Y DE...,<NA>,1,Activo,1983-10-25,2026-08-05,True
70893,26583837,1057066,0,295436,VIRGEN DE FATIMA DE BARRANCO,D2,Básica Alternativa-Avanzado,No aplica,3,Mixto,...,11,Mañana,26162873262,MISIONERAS SEGLARES DE JESUS SACRAMENTADO Y DE...,<NA>,1,Activo,1993-01-10,2026-08-05,True


Correos con formato inválido:


,CODINST,COD_MOD,ANEXO,CODLOCAL,CEN_EDU,NIV_MOD,D_NIV_MOD,D_FORMA,TIPSSEXO,D_TIPSSEXO,...,COD_TUR,D_COD_TUR,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo
4699,,1117662,0,068448,SAN PEDRO,L0,Técnico Productiva,Escolarizada,3,Mixto,...,12,Tarde,<NA>,,GEORGE SCERRI ABELA,1,Activo,2004-05-14,2026-08-05,True
6091,,1334408,0,546920,SYSTEM MILENIUM,L0,Técnico Productiva,Escolarizada,3,Mixto,...,13,Mañana-Tarde,<NA>,,<NA>,2,Inactivo,2004-01-01,2026-08-05,False
17554,,0927335,0,717342,KHIPU,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,2,Inactivo,1993-08-25,2026-08-05,False
17574,26914345,0928358,0,147083,PUKLLASUNCHIS,F0,Secundaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1993-07-21,2026-08-05,True
18136,21691917,0730523,0,169715,166 VIRGEN DEL CARMEN,A2,Inicial - Jardín,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1987-04-23,2026-08-05,True
33818,,0240135,0,441782,PUNO,M0,Escuela Formación Artística,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1986-01-09,2026-08-05,True
39050,26970768,0929133,0,042750,INDIVISA MANENT DE LA SALLE,F0,Secundaria,Escolarizada,3,Mixto,...,11,Mañana,20484849723,ASOC.CIVIL DE SERV.EDUCAT.INDIVISA MANEN,<NA>,1,Activo,NaT,2026-08-05,True
47476,20822791,0222182,0,127061,16462 SAN JUAN BOSCO,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,1990-06-20,2026-08-05,True
48030,,0594721,0,590828,ARZOBISPO OSCAR ROMERO,L0,Técnico Productiva,Escolarizada,3,Mixto,...,16,Mañana-Noche,<NA>,,<NA>,1,Activo,NaT,2026-08-05,True
54707,25286936,0910562,0,314920,7264 DR MANUEL MIGUEL VALLE Y VALLE,B0,Primaria,Escolarizada,3,Mixto,...,11,Mañana,<NA>,,<NA>,1,Activo,NaT,2026-08-05,True


### <font color=skyblue> 3.6 Tratamiento básico de valores inválidos y faltantes — Padron_web </font>### <font color=green>

In [ ]:
#padron_web
# Convertir valores inválidos a NaN
padron.loc[padron["FECHAREG"] > pd.Timestamp("today").normalize(), "FECHAREG"] = pd.NaT
padron.loc[padron["FECHA_ACT"] < padron["FECHAREG"], "FECHAREG"] = pd.NaT

padron.loc[padron["NRORUC"].notna() &
           ~padron["NRORUC"].str.match(r"^(10|15|16|17|20)\d{9}$").fillna(False), "NRORUC"] = pd.NA
padron.loc[padron["EMAIL"].notna() &
           ~padron["EMAIL"].str.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$").fillna(False), "EMAIL"] = pd.NA

# Imputación de coordenadas
mediana_distrito = padron.groupby("CODGEO")[["NLAT_IE", "NLONG_IE"]].transform("median")
padron["coord_imputada"] = padron["NLAT_IE"].isna() & mediana_distrito["NLAT_IE"].notna()
padron["NLAT_IE"]  = padron["NLAT_IE"].fillna(mediana_distrito["NLAT_IE"])
padron["NLONG_IE"] = padron["NLONG_IE"].fillna(mediana_distrito["NLONG_IE"])

# Imputación simple
padron["TELEFONO"]   = padron["TELEFONO"].fillna("No especificado")
padron["EMAIL"]      = padron["EMAIL"].fillna("No especificado")
padron["PAGWEB"]     = padron["PAGWEB"].fillna("No especificado")
padron["DIRECTOR"]   = padron["DIRECTOR"].fillna("No especificado")
padron["LOCALIDAD"]  = padron["LOCALIDAD"].fillna("No especificado")
padron["CEN_POB"]    = padron["CEN_POB"].fillna("No especificado")
padron["D_TIPOPROG"] = padron["D_TIPOPROG"].fillna("No aplica")   # no es programa no escolarizado

print("Faltantes después del tratamiento:")
display(padron.isna().sum())

/tmp/ipykernel_3695/2481255527.py:7: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  ~padron["NRORUC"].str.match(r"^(10|15|16|17|20)\d{9}$").fillna(False), "NRORUC"] = pd.NA


Faltantes después del tratamiento:


,0
CODINST,0
COD_MOD,0
ANEXO,0
CODLOCAL,0
CEN_EDU,0
NIV_MOD,0
D_NIV_MOD,0
D_FORMA,0
TIPSSEXO,0
D_TIPSSEXO,0


### <font color=skyblue> 3.6 Eliminación o consolidación de duplicados — Padron_web </font>

In [ ]:
# "censo" es el dataset ya limpio de la sección de Lineal_3AP (F1), no una fuente aparte
censo = df_censo_final_limpio.copy()
padron_limpio = padron.drop_duplicates(subset=["COD_MOD", "ANEXO"], keep="first")

print("Padrón antes y después:", padron.shape, padron_limpio.shape)
print("Censo 3AP antes y después:", censo.shape, censo.shape)  # censo ya viene deduplicado

Padrón antes y después: (180792, 47) (180792, 47)
Censo 3AP antes y después: (37354, 41) (37354, 41)


In [ ]:
padron_limpio.dtypes

,0
CODINST,object
COD_MOD,object
ANEXO,object
CODLOCAL,object
CEN_EDU,object
NIV_MOD,object
D_NIV_MOD,string[python]
D_FORMA,string[python]
TIPSSEXO,object
D_TIPSSEXO,string[python]


## <font color=yellow> 4. Integración de datos</font>

Integramos las tres fuentes en dos pasos, porque no comparten la misma unidad de análisis:

- `df_censo_final_limpio` y `padron_limpio` sí son ambas a nivel **colegio** (llave `COD_MOD` + `ANEXO`), así que se integran con un merge 1 a 1.
- `df2_limpio` (denuncias) es a nivel **caso de denuncia**, sin identificador de colegio. La llave común más fina que comparte con las otras dos fuentes es la **DRE**, así que la agregamos (conteo de denuncias por DRE) antes de unirla a la base escolar.

### <font color=skyblue> 4.2 Integración básica </font>

In [ ]:
base_censo = pd.merge(
    df_censo_final,
    padron_limpio,
    on=["COD_MOD", "ANEXO"],
    how="outer",
    suffixes = ("_censo","_padron"),
    indicator = True
)

print("Base escolar integrada con llave compuesta (COD_MOD + ANEXO):", base_censo.shape)
display(base_censo.head())

Base escolar integrada con llave compuesta (COD_MOD + ANEXO): (180792, 87)


,COD_MOD,ANEXO,CODLOCAL_censo,NROCED,CEN_EDU_censo,DISTRITO,LOCALIDAD_censo,CODGEO_censo,NIV_MOD_censo,GES_DEP_censo,...,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo,coord_imputada,_merge
0,0000701,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-01-10,2026-08-05,False,True,right_only
1,0000702,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-03-22,2026-08-05,False,False,right_only
2,0000704,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2016-01-27,2026-08-05,False,False,right_only
3,0001505,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-03-22,2026-08-05,False,False,right_only
4,0001506,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-03-22,2026-08-05,False,False,right_only


In [ ]:
base_censo = base_censo.rename(columns={'D_DREUGEL': 'UGEL_norm'})
base_censo.head(5)

,COD_MOD,ANEXO,CODLOCAL_censo,NROCED,CEN_EDU_censo,DISTRITO,LOCALIDAD_censo,CODGEO_censo,NIV_MOD_censo,GES_DEP_censo,...,NRORUC,RZSOCIAL,PROMOTOR,ESTADO,D_ESTADO,FECHAREG,FECHA_ACT,activo,coord_imputada,_merge
0,0000701,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-01-10,2026-08-05,False,True,right_only
1,0000702,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-03-22,2026-08-05,False,False,right_only
2,0000704,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2016-01-27,2026-08-05,False,False,right_only
3,0001505,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-03-22,2026-08-05,False,False,right_only
4,0001506,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,<NA>,,<NA>,2,Inactivo,2012-03-22,2026-08-05,False,False,right_only


In [ ]:
columnas_interes = ['COD_MOD', 'CEN_EDU_censo', 'DISTRITO', 'UGEL_norm', 'D_REGION']
display(base_censo[columnas_interes].head(10))

,COD_MOD,CEN_EDU_censo,DISTRITO,UGEL_norm,D_REGION
0,0000701,NaN,NaN,UGEL LA UNION,DRE AREQUIPA
1,0000702,NaN,NaN,UGEL LA UNION,DRE AREQUIPA
2,0000704,NaN,NaN,UGEL INTERCULTURAL BILINGUE CONDORCANQUI,DRE AMAZONAS
3,0001505,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS
4,0001506,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS
5,0001507,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS
6,0001509,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS
7,0001510,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS
8,0001512,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS
9,0001513,NaN,NaN,UGEL 13 YAUYOS,DRE LIMA PROVINCIAS


In [ ]:
conteo_cruce = base_censo["_merge"].value_counts().reset_index() #cuenta cuántas filas hay de cada tipo de merge
conteo_cruce.columns = ["resultado_cruce", "cantidad"]
conteo_cruce

,resultado_cruce,cantidad
0,right_only,143438
1,both,37354
2,left_only,0


In [ ]:
df_censo_final_limpio.shape

(37354, 41)

In [ ]:
base_censo.dtypes

,0
COD_MOD,object
ANEXO,object
CODLOCAL_censo,object
NROCED,object
CEN_EDU_censo,object
...,...
FECHAREG,datetime64[ns]
FECHA_ACT,datetime64[ns]
activo,boolean
coord_imputada,bool


In [ ]:
base_censo_final = pd.merge(
    base_censo,
    df_violencia_por_ugel,
    on=["UGEL_norm"],
    how="outer",
    suffixes = ("_primer_censo","_denuncia_por_ugel"),
    indicator = "merge_denuncias" # Se cambia True por un nombre único
)

print("Base escolar integrada final:", base_censo_final.shape)
display(base_censo_final.head())

Base escolar integrada final: (180841, 93)


,COD_MOD,ANEXO,CODLOCAL_censo,NROCED,CEN_EDU_censo,DISTRITO,LOCALIDAD_censo,CODGEO_censo,NIV_MOD_censo,GES_DEP_censo,...,FECHA_ACT,activo,coord_imputada,_merge,nro_denuncias_tipo_fisica,nro_denuncias_tipo_psicologica,nro_denuncias_tipo_sexual,nro_denuncias_tipo_reporte_entreescolares,nro_denuncias_tipo_reporte_personalescolar,merge_denuncias
0,0359729,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
1,0412189,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
2,0413518,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
3,0416123,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
4,0543355,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,False,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only


### <font color=skyblue> 4.1 Eliminación de columnas no necesarias </font>

In [ ]:
# Lista de columnas redundantes a eliminar (se prefiere mantener la versión _censo o la original)
columnas_a_eliminar = [
    'ANEXO_padron',
    'CODLOCAL_padron',
    'CEN_EDU_padron',
    'NIV_MOD_padron',
    'GES_DEP_padron',
    'LOCALIDAD_padron'
]

# Eliminar las columnas redundantes de base_censo_final
base_censo_final = base_censo_final.drop(columns=columnas_a_eliminar, errors='ignore')

# Renombrar las columnas restantes para quitarles el sufijo si es necesario
base_censo_final = base_censo_final.rename(columns={
    'ANEXO_censo': 'ANEXO',
    'CODLOCAL_censo': 'CODLOCAL',
    'CEN_EDU_censo': 'CEN_EDU',
    'NIV_MOD_censo': 'NIV_MOD',
    'GES_DEP_censo': 'GES_DEP',
    'LOCALIDAD_censo': 'LOCALIDAD'
})

print(f"Columnas eliminadas: {columnas_a_eliminar}")
print(f"Nuevas dimensiones de base_censo_final: {base_censo_final.shape}")
display(base_censo_final.head())


Columnas eliminadas: ['ANEXO_padron', 'CODLOCAL_padron', 'CEN_EDU_padron', 'NIV_MOD_padron', 'GES_DEP_padron', 'LOCALIDAD_padron']
Nuevas dimensiones de base_censo_final: (180841, 88)


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO_censo,NIV_MOD,GES_DEP,...,FECHA_ACT,activo,coord_imputada,_merge,nro_denuncias_tipo_fisica,nro_denuncias_tipo_psicologica,nro_denuncias_tipo_sexual,nro_denuncias_tipo_reporte_entreescolares,nro_denuncias_tipo_reporte_personalescolar,merge_denuncias
0,0359729,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
1,0412189,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
2,0413518,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
3,0416123,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,True,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only
4,0543355,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2026-08-05,False,False,right_only,NaN,NaN,NaN,NaN,NaN,left_only


In [ ]:
# Eliminar filas con más de 45 valores NaN
# Threshold = Total de columnas (94) - Máximo de nulos permitido (45) = 49
base_censo_final_filtrada = base_censo_final.dropna(thresh=49)

print(f"Dimensiones originales: {base_censo_final.shape}")
print(f"Dimensiones tras filtrar filas con más de 45 NaNs: {base_censo_final_filtrada.shape}")
print(f"Filas eliminadas: {len(base_censo_final) - len(base_censo_final_filtrada)}")

display(base_censo_final_filtrada.head())

Dimensiones originales: (180841, 88)
Dimensiones tras filtrar filas con más de 45 NaNs: (37422, 88)
Filas eliminadas: 143419


,COD_MOD,ANEXO,CODLOCAL,NROCED,CEN_EDU,DISTRITO,LOCALIDAD,CODGEO_censo,NIV_MOD,GES_DEP,...,FECHA_ACT,activo,coord_imputada,_merge,nro_denuncias_tipo_fisica,nro_denuncias_tipo_psicologica,nro_denuncias_tipo_sexual,nro_denuncias_tipo_reporte_entreescolares,nro_denuncias_tipo_reporte_personalescolar,merge_denuncias
244,0207795,0,139766,3AP,5040 Pedro Ruiz,Callao,Callao,070101,B0,A1,...,2026-08-05,True,False,both,485.0,510.0,160.0,806.0,349.0,both
245,0207803,0,143438,3AP,5041 Nuestra Seðora Del Monte Carmelo,Carmen De La Legua Reynoso,Callao,070103,B0,A1,...,2026-08-05,True,False,both,485.0,510.0,160.0,806.0,349.0,both
246,0207845,0,139870,3AP,5031 Cesar Vallejo,Callao,Manuel Cipriano Dulanto Valenzuela,070101,B0,A1,...,2026-08-05,True,False,both,485.0,510.0,160.0,806.0,349.0,both
247,0207852,0,139827,3AP,5032 Enrique Delhorme,Callao,Tarapac┴,070101,B0,A1,...,2026-08-05,True,False,both,485.0,510.0,160.0,806.0,349.0,both
248,0207894,0,139889,3AP,5019 Augusto Cazorla,Callao,Callao,070101,B0,A1,...,2026-08-05,True,False,both,485.0,510.0,160.0,806.0,349.0,both


## <font color=yellow> 5. Visualización con gráficos</font>

### <font color=skyblue> 5.1 Gráfico 1 - Barras </font>

In [ ]:
import plotly.express as px

# Agrupar denuncias por departamento
# Usamos base_censo_final que contiene las columnas de denuncias y D_DPTO
denuncias_por_dpto = base_censo_final.groupby('D_DPTO')[['nro_denuncias_tipo_fisica', 'nro_denuncias_tipo_psicologica', 'nro_denuncias_tipo_sexual']].sum().reset_index()

# Transformar el dataframe para que sea compatible con un gráfico de barras apilado o agrupado
data_plot = denuncias_por_dpto.melt(id_vars='D_DPTO',
                                    value_vars=['nro_denuncias_tipo_fisica', 'nro_denuncias_tipo_psicologica', 'nro_denuncias_tipo_sexual'],
                                    var_name='Tipo de Violencia',
                                    value_name='Cantidad')

# Limpiar los nombres de las categorías para la leyenda
data_plot['Tipo de Violencia'] = data_plot['Tipo de Violencia'].str.replace('nro_denuncias_tipo_', '').str.capitalize()

# Crear el gráfico de barras horizontal
fig = px.bar(
    data_plot,
    y="D_DPTO",
    x="Cantidad",
    color="Tipo de Violencia",
    orientation='h',
    title="Cantidad de denuncias por tipo de violencia y departamento",
    labels={'D_DPTO': 'Departamento', 'Cantidad': 'Número de Denuncias'},
    barmode='group'
)

fig.update_layout(yaxis={'categoryorder':'total ascending'}, height=800)
fig.show()


**Conclusión**

2. Prevalencia de Tipos: En la mayoría de departamentos, la violencia psicológica y física superan significativamente a la violencia sexual en número de reportes. Es decir, la violencia más predominante en el Perú es la psicologica seguida por la física.

3. Brechas Regionales: Existe una diferencia drástica entre los departamentos con más reportes (Lima, Piura, Arequipa) y aquellos con menores cifras, lo que sugiere la necesidad de fortalecer los canales de denuncia en regiones con menor registro porque puede que no denuncias por falta de canales de denuncia

### <font color=skyblue> 5.2 Gráfico 2 - Barras </font>

In [ ]:
import plotly.express as px

# Agrupamos por tipo de gestión y sumamos las denuncias principales
# D_GESTION indica si es Público o Privado
resumen_gestion = data_lima_metro.groupby('D_GESTION')[['nro_denuncias_tipo_psicologica', 'nro_denuncias_tipo_fisica']].mean().reset_index()

data_plot_gestion = resumen_gestion.melt(id_vars='D_GESTION',
                                         var_name='Tipo de Denuncia',
                                         value_name='Promedio de Denuncias por Colegio')

# Limpiar etiquetas
data_plot_gestion['Tipo de Denuncia'] = data_plot_gestion['Tipo de Denuncia'].str.replace('nro_denuncias_tipo_', '').str.capitalize()

fig_gestion = px.bar(
    data_plot_gestion,
    x='D_GESTION',
    y='Promedio de Denuncias por Colegio',
    color='Tipo de Denuncia',
    barmode='group',
    title='Promedio de Denuncias (Psicológica vs Física) según Gestión en Lima Metropolitana',
    labels={'D_GESTION': 'Tipo de Gestión', 'Promedio de Denuncias por Colegio': 'Promedio de Casos'},
    template='plotly_white'
)

fig_gestion.show()

**CONCLUSIÓN**
1. El tipo de violencia predominante es la violencia psicológica en todos los distintos niveles de gestión de los colegios (privada, pública de gestión directa, pública de gestión privada).

### <font color=skyblue> 5.3 Gráfico 3 - Diagrama de caja </font>

Comparación de Distribución: Gestión Pública vs. Privada
Utilizamos un Box Plot para visualizar la dispersión de denuncias y detectar valores atípicos en Lima Metropolitana.

In [ ]:
import plotly.express as px

# Corregimos el warning usando .loc para la asignación
data_lima_metro = data_lima_metro.copy()
data_lima_metro.loc[:, 'total_denuncias'] = data_lima_metro['nro_denuncias_tipo_psicologica'] + data_lima_metro['nro_denuncias_tipo_fisica']

# 1. Visualización de distribución
fig_box = px.box(
    data_lima_metro,
    x='D_GESTION',    y='total_denuncias',
    color='D_GESTION',
    notched=True,
    title='Análisis de Distribución: Gestión Pública vs Privada (Lima Metropolitana)',
    labels={'D_GESTION': 'Gestión', 'total_denuncias': 'Total Denuncias'},
    template='plotly_white'
)

fig_box.show()

# 2. Resumen estadístico detallado
resumen_stats = data_lima_metro.groupby('D_GESTION')['total_denuncias'].agg(['mean', 'median', 'std', 'max', 'count']).reset_index()
print("Validación Estadística final:")
display(resumen_stats)


Validación Estadística final:


,D_GESTION,mean,median,std,max,count
0,Privada,1476.765991,1465.0,376.472324,2269.0,2517
1,Pública de gestión directa,1472.695518,1465.0,370.314491,2269.0,647
2,Pública de gestión privada,1448.621622,1465.0,371.949993,2269.0,37


## <font color=yellow> 6. Conclusión </font>
1. Mayor Variabilidad en Gestión Pública: La desviación estándar y el rango intercuartílico son más amplios en los colegios públicos, lo que indica una mayor heterogeneidad en la incidencia o reporte de violencia.
  
3. Presencia de Outliers Críticos: Existen valores atípicos extremos (máximos muy alejados de la media) en ambos sectores, pero el volumen de casos en los puntos más altos es mayor en la gestión pública.

